platform/wayland/cog-utils-wl.c

```diff
@@ -187,6 +187,180 @@ cog_wl_display_destroy(CogWlDisplay *display)
     g_slice_free(CogWlDisplay, display);
 }
 
+static void
+xdg_popup_on_configure(void *data, struct xdg_popup *xdg_popup, int32_t x, int32_t y, int32_t width, int32_t height)
+{
+}
+
+static void
+xdg_popup_on_popup_done(void *data, struct xdg_popup *xdg_popup)
+{
+    cog_wl_popup_destroy((CogWlPopup *) data);
+}
+
+static void
+shell_popup_surface_ping(void *data, struct wl_shell_surface *shell_surface, uint32_t serial)
+{
+    wl_shell_surface_pong(shell_surface, serial);
+}
+
+static void
+shell_popup_surface_configure(void                    *data,
+                              struct wl_shell_surface *shell_surface,
+                              uint32_t                 edges,
+                              int32_t                  width,
+                              int32_t                  height)
+{
+}
+
+static void
+shell_popup_surface_popup_done(void *data, struct wl_shell_surface *shell_surface)
+{
+}
+
+static const struct wl_shell_surface_listener shell_popup_surface_listener = {
+    .ping = shell_popup_surface_ping,
+    .configure = shell_popup_surface_configure,
+    .popup_done = shell_popup_surface_popup_done,
+};
+
+static void
+xdg_surface_on_configure(void *data, struct xdg_surface *surface, uint32_t serial)
+{
+    xdg_surface_ack_configure(surface, serial);
+
+    CogWlPopup *popup = (CogWlPopup *) data;
```

Comment:
FWIW, casting from `void*` to a concrete pointer type does not need the cast, in C the following is valid and preferred:

```suggestion
    CogWlPopup *popup = data;
```